kvserver: add kv.closed_timestamp.lead_for_global_reads_auto_tune_interval

wenyihu6 opened the pull request:
**kvserver: add kv.closed_timestamp.lead_for_global_reads_auto_tune_interval**

This commit adds a new cluster setting,
`kv.closed_timestamp.lead_for_global_reads_auto_tune_interval` to control how
frequently the system measures latency between nodes. If non-zero, the observed
latency between nodes in the cluster would be used to dynamically auto-tune
the lead time for ranges using the LEAD_FOR_GLOBAL_READS policy.
The setting defaults to 5 minutes and can be disabled by setting to 0.
It only takes effect when `side_transport_interval` is non-zero.

Note that setting is currently unused. Future commits will populate and
integrate it.

Part of: #59680
Release note: none

---

**kvserver/closedts: use max() instead of conditional checks**

Previously, conditional check is used to find the maximum
between sideTransportPropTime and raftTransportPropTime.

This patch changes it to use the max builtin for better
readability.

Release note: none
Epic: none

---

**kvserver/closedts: move leadTargetOverride above for better readability**

Previously, if a cluster defines `kv.closed_timestamp.lead_for_global_reads_override`,
the lead time calclulated at the sender will just be overriden in the end.

This patch moves the override part earlier in the logic and
breaks on an early return for better readability.

Epic: none
Release note: none

---

**kvserver: move closed ts lead time computation into a helper function**

This patch extracts the calculation of closed timestamp lead time for ranges
configured with global reads policies to a helper function that takes in
networkRTT as a parameter. Currently, it always takes in
closedts.DefaultMaxNetworkRTT. But future commits will change it to take in
a dynamically observed network latency based on policies.

Epic: none
Release note: none

---

**kvserver/closedts: clean up TargetForPolicy**

This commit cleans up the closedts.TargetForPolicy logic a
bit to simplify future commits.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: use max() instead of conditional checks

Previously, conditional check was used to find the maximum
between sideTransportPropTime and raftTransportPropTime.
This commit changes it to use the max builtin for better
readability.

Release note: none
Epic: none

wenyihu6 committed:
kvserver/closedts: move leadTargetOverride above for better readability

Previously, if a cluster defines `kv.closed_timestamp.lead_for_global_reads_override`,
the lead time calclulated at the sender will just be overriden in the end.

This patch moves the override part earlier in the logic and
breaks on an early return for better readability.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: add kv.closed_timestamp.lead_for_global_reads_auto_tune_interval

This commit adds a new cluster setting,
`kv.closed_timestamp.lead_for_global_reads_auto_tune_interval` to control how
frequently the system measures latency between nodes. If non-zero, the observed
latency between nodes in the cluster would be used to dynamically auto-tune
the lead time for ranges using the LEAD_FOR_GLOBAL_READS policy.
The setting defaults to 5 minutes and can be disabled by setting to 0.
It only takes effect when `side_transport_interval` is non-zero.

Note that setting is currently unused. Future commits will populate and
integrate it.

Part of: #59680
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142281)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver: move closed ts lead time computation into a helper function

This patch extracts the calculation of closed timestamp lead time for ranges
configured with global reads policies to a helper function that takes in
networkRTT as a parameter. Currently, it always takes in
closedts.DefaultMaxNetworkRTT. But future commits will change it to take in
a dynamically observed network latency based on policies.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: clean up TargetForPolicy

This commit cleans up the closedts.TargetForPolicy logic a
bit to simplify future commits.

Epic: none
Release note: none

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo comments.

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142281#-:-OM8WFAJ8KaadobiOir0:b46bmkf)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`-- commits` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142281#-OM8UijJ2nGtDYvjgc-I:-OM8UijJ2nGtDYvjgc-J:b5il6pq):*
nit: s/is used/was used/g
___
*[`-- commits` line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142281#-OM8UlyW2xWuKbHpL8IL:-OM8UlyX3Lp0aEWV-HDR:bow72k):*
nit: this could just be appended to the previous paragraph.
___
*[`pkg/kv/kvserver/closedts/policy.go` line 109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142281#-OM8W-Xu15ujqe5IFrh7:-OM8W-Xu15ujqe5IFrh8:bn69pit) ([raw file](https://github.com/cockroachdb/cockroach/blob/62d0d1233199cd107594a6e82121a2dba49925f9/pkg/kv/kvserver/closedts/policy.go#L109)):*
> ```Go
> 	policy roachpb.RangeClosedTimestampPolicy,
> ) hlc.Timestamp {
> 	var leadTimeAtSender time.Duration
> ```

`leadTimeAtSender` doesn't seem like the right variable name when dealing with the `LAG_BY_CLUSTER_SETTING` policy. Let's either call this something generic, like `delta`, or drop this commit entirely.
___
*[`pkg/kv/kvserver/closedts/setting.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142281#-OM8UZnHDonPdtfh43Nl:-OM8UZnHDonPdtfh43Nm:b-ayg9vz) ([raw file](https://github.com/cockroachdb/cockroach/blob/065aaf82806a21622868b5b61ed4622b0919f4e7/pkg/kv/kvserver/closedts/setting.go#L53)):*
> ```Go
> )
> 
> // LeadForGlobalReadsAutoTuneInterval controls how frequently the system adjusts closed
> ```

nit: let's wrap comments to 80 chars


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142281)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani)

___
*[`-- commits` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142281#-OM8UijJ2nGtDYvjgc-I:-OM8cbnG5-ArvJdbwNcA:b-896fix):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/is used/was used/g
</blockquote></details>

Done.
___
*[`-- commits` line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142281#-OM8UlyW2xWuKbHpL8IL:-OM8cdfm-3EIzxfIHGuE:b-896fix):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: this could just be appended to the previous paragraph.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/closedts/policy.go` line 109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142281#-OM8W-Xu15ujqe5IFrh7:-OM8cez-1Vhg7irJvhZU:b5djafk) ([raw file](https://github.com/cockroachdb/cockroach/blob/62d0d1233199cd107594a6e82121a2dba49925f9/pkg/kv/kvserver/closedts/policy.go#L109)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

`leadTimeAtSender` doesn't seem like the right variable name when dealing with the `LAG_BY_CLUSTER_SETTING` policy. Let's either call this something generic, like `delta`, or drop this commit entirely.
</blockquote></details>

Renamed it to `targetOffsetTime`.
___
*[`pkg/kv/kvserver/closedts/setting.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142281#-OM8UZnHDonPdtfh43Nl:-OM8ciuR07-1gWupepWw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/065aaf82806a21622868b5b61ed4622b0919f4e7/pkg/kv/kvserver/closedts/setting.go#L53)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: let's wrap comments to 80 chars
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR!

bors r=arulajmani
